mlflow/pyfunc/scoring_server/client.py

```diff
@@ -49,31 +73,49 @@ def invoke(self, data):
         Invoke inference on input data. The input data must be pandas dataframe or numpy array or
         a dict of numpy arrays.
         """
-        content_type = scoring_server.CONTENT_TYPE_JSON
-
-        def get_jsonable_input(name, data):
-            if isinstance(data, np.ndarray):
-                return data.tolist()
-            else:
-                raise MlflowException(f"Incompatible input type:{type(data)} for input {name}.")
-
-        if isinstance(data, pd.DataFrame):
-            post_data = {"dataframe_split": data.to_dict(orient="split")}
-        elif isinstance(data, dict):
-            post_data = {"inputs": {k: get_jsonable_input(k, v) for k, v in data}}
-        elif isinstance(data, np.ndarray):
-            post_data = ({"inputs": data.tolist()},)
-        else:
-            post_data = data
-        if not isinstance(post_data, str):
-            post_data = json.dumps(post_data, cls=_CustomJsonEncoder)
         response = requests.post(
             url=self.url_prefix + "/invocations",
-            data=post_data,
-            headers={"Content-Type": content_type},
+            data=dump_data(data),
+            headers={"Content-Type": scoring_server.CONTENT_TYPE_JSON},
         )
         if response.status_code != 200:
             raise Exception(
                 f"Invocation failed (error code {response.status_code}, response: {response.text})"
             )
         return PredictionsResponse.from_json(response.text)
+
+
+class StdinScoringServerClient:
+    def __init__(self, process):
+        self.process = process
+
+    def wait_until_file_exists(self, f):
+        begin_time = time.time()
+        while True:
+            if f.exists():
+                return
+            time.sleep(1)
+            if time.time() - begin_time > 30:
+                raise RuntimeError("Timeout")
+
+    def invoke(self, data):
+        """
+        Invoke inference on input data. The input data must be pandas dataframe or numpy array or
+        a dict of numpy arrays.
+        """
+        with tempfile.TemporaryDirectory() as tmpdir:
+            tmpdir = Path(tmpdir)
+            output_file = tmpdir.joinpath(f"{uuid.uuid4()}.json")
```

Comment:
Creating a new temporary file for every invocation could get very expensive.

Could we instead use one temporary file per `StdinScoringServerClient`? I think we can do this if we generate a UUID for each request and then add the results to the file with the UUID, e.g. write the following to the file:

```
{
    "request_id": <uuid>,
    "response": <the model response JSON>
}
```

Reply:
Sounds good! I'll try what you suggested.